File: app01_java_nft/backend/src/main/java/com/devpowers/nftcerts/api/UploadController.java
package com.devpowers.nftcerts.api;

import com.devpowers.nftcerts.error.EmptyFileException;
import com.devpowers.nftcerts.error.UnsupportedFileTypeException;
import com.devpowers.nftcerts.hashing.Sha256HashingService;
import com.devpowers.nftcerts.watermark.MetadataWatermarkService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Set;
import java.util.UUID;

/**
 * POST /api/uploads — accepts a photograph, computes its SHA-256 "digital fingerprint", embeds a
 * best-effort EXIF watermark, and stores the (possibly watermarked) bytes for later use when
 * creating an artwork.
 */
@RestController
public class UploadController {

    private static final Set<String> SUPPORTED_CONTENT_TYPES = Set.of("image/jpeg", "image/png", "image/webp");

    private final Sha256HashingService hashingService;
    private final MetadataWatermarkService watermarkService;
    private final UploadStore uploadStore;

    public UploadController(Sha256HashingService hashingService, MetadataWatermarkService watermarkService,
                             UploadStore uploadStore) {
        this.hashingService = hashingService;
        this.watermarkService = watermarkService;
        this.uploadStore = uploadStore;
    }

    @PostMapping(value = "/api/uploads", consumes = "multipart/form-data")
    public ResponseEntity<UploadResponse> upload(@RequestParam("file") MultipartFile file) {
        if (file.isEmpty()) {
            throw new EmptyFileException();
        }
        String contentType = file.getContentType();
        if (contentType == null || !SUPPORTED_CONTENT_TYPES.contains(contentType.toLowerCase())) {
            throw new UnsupportedFileTypeException(contentType);
        }

        byte[] originalContent;
        try {
            originalContent = file.getBytes();
        } catch (IOException e) {
            throw new UncheckedIOException("Failed to read uploaded file", e);
        }

        String sha256Hash = hashingService.sha256Hex(originalContent);
        // Watermark payload references the hash of the *original* content; the artist DID is not
        // yet known at upload time (it is supplied later on /api/artworks), so an empty marker is
        // used here and the definitive DID-linked watermark/hash pairing lives in the metadata JSON
        // pinned to IPFS during artwork creation.
        byte[] watermarkedContent = watermarkService.watermark(originalContent, contentType, sha256Hash, "");

        StoredUpload storedUpload = new StoredUpload(file.getOriginalFilename(), contentType, sha256Hash, watermarkedContent);
        UUID fileId = uploadStore.store(storedUpload);

        UploadResponse response = new UploadResponse(fileId, file.getOriginalFilename(), sha256Hash, watermarkedContent.length);
        return ResponseEntity.ok(response);
    }
}

File: app02_cpp_nft/backend/src/api/UploadController.cpp
#include "api/UploadController.h"

#include "error/Exceptions.h"

#include <drogon/HttpAppFramework.h>
#include <drogon/MultiPart.h>
#include <json/json.h>

#include <algorithm>
#include <set>

namespace nftcerts::api {

namespace {

std::string toLower(const std::string& s) {
    std::string result = s;
    std::transform(result.begin(), result.end(), result.begin(), ::tolower);
    return result;
}

// Drogon's HttpFile::getContentType() returns an internal enum rather than a raw MIME string, so
// content type is derived from the file extension instead — equivalent in practice for this
// reference implementation's fixed supported set (see docs/sdlc/05-api-design.md #1).
std::string contentTypeFromExtension(const std::string& extension) {
    std::string ext = toLower(extension);
    if (ext == "jpg" || ext == "jpeg") return "image/jpeg";
    if (ext == "png") return "image/png";
    if (ext == "webp") return "image/webp";
    return "";
}

}  // namespace

void registerUploadRoutes(hashing::Sha256HashingService& hashingService,
                           watermark::MetadataWatermarkService& watermarkService, UploadStore& uploadStore) {
    drogon::app().registerHandler(
        "/api/uploads",
        [&hashingService, &watermarkService, &uploadStore](
            const drogon::HttpRequestPtr& req, std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
            drogon::MultiPartParser parser;
            if (parser.parse(req) != 0 || parser.getFiles().empty()) {
                throw error::EmptyFileException();
            }

            const drogon::HttpFile& file = parser.getFiles()[0];
            if (file.fileLength() == 0) {
                throw error::EmptyFileException();
            }

            std::string contentType = contentTypeFromExtension(std::string(file.getFileExtension()));
            if (contentType.empty()) {
                throw error::UnsupportedFileTypeException(std::string(file.getFileExtension()));
            }

            std::vector<unsigned char> originalContent(file.fileData(), file.fileData() + file.fileLength());

            std::string sha256Hash = hashingService.sha256Hex(originalContent);
            // Watermark payload references the hash of the *original* content; the artist DID is
            // not yet known at upload time (supplied later on /api/artworks), so an empty marker
            // is used here — mirrors app01's UploadController.
            std::vector<unsigned char> watermarkedContent =
                watermarkService.watermark(originalContent, contentType, sha256Hash, "");

            certificate::StoredUpload stored{file.getFileName(), contentType, sha256Hash, watermarkedContent};
            std::string fileId = uploadStore.store(stored);

            Json::Value response;
            response["fileId"] = fileId;
            response["originalFilename"] = file.getFileName();
            response["sha256Hash"] = sha256Hash;
            response["sizeBytes"] = static_cast<Json::UInt64>(watermarkedContent.size());

            callback(drogon::HttpResponse::newHttpJsonResponse(response));
        },
        {drogon::Post});
}

}  // namespace nftcerts::api
